src/Runtime/Core/Serialization/InferDataContractResolver.cs

```diff
@@ -39,11 +154,428 @@ public override Type ResolveName(string typeName, string typeNamespace, Type dec
             var ty = knownTypeResolver.ResolveName(typeName, typeNamespace, declaredType, knownTypeResolver);
 
             if (ty != null) return ty;
+            
+            if (!typeNamespace.StartsWith(nsPrefix)) // Is this our namespace?
+            {
+                return null; // Cannot resolve.
+            }
+
+            // We limit the length of the type name we are willing to try to parse
+            // for security reasons.
+            if (typeName.Length > 10000)
+            {
+                throw new Exception("Type name exceeded maximum length");
+            }
+
+            return ConstructTypeFromString(typeName);
+        }
+
+        internal static Type ConstructTypeFromString(string typeString)
+        {
+            // We avoid loading a type from a string for security reasons before we
+            // first parse it to ensure that it is constructed from approved types.
+            var typeLayout = GetTypeLayout(typeString);
+
+            // We now construct the types entirely from the allowed list, using the 
+            // type layout to direct how we combine types in the allowed list.
+            // This means that even if the data is tampered with, a type is never
+            // loaded -- types are only constructed as combinations of types in the
+            // allowed list.
+            return ConstructTypeFromAllowedlist(typeLayout);
+        }
+
+        /// <summary>
+        /// Some information about the layout of a type.
+        /// </summary>
+        internal sealed class TypeId
```

Comment:
`TypeId` is not a good name since this is already used by [Attribute](https://learn.microsoft.com/en-us/dotnet/api/system.attribute.typeid?view=net-7.0) to mean something else.  [TypeLayout](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.typelayout?view=net-7.0) is also used.  What this really represents is a `ParsedTypeString`.

Reply:
Renamed.